src/main/java/org/openjdk/shenandoah/Snapshot.java

```diff
@@ -260,4 +279,148 @@ public double percentageOfOldRegionsInCollectionSet() {
                 old_in_cset, total_in_cset, old, total);
         return total_in_cset == 0 ? 0 : ((double) (old_in_cset)) / total_in_cset;
     }
+    public int emptyUncommittedCounter() {
+        for (RegionStat rs : stats) {
```

Comment:
Does each Region in the heap have a corresponding RegionStat?

If so, wouldn't it be better to sweep over the regions once and update/accumulate all of the counters so they can be used when needed? Something like

```public updateAllCountersForRegions() {
  for (RegionStat rs: stats) {
       switch (rs.state()) {
          case RegionState.EMPTY_COMMITTED: 
               counters[EMPTY_COMMITTED]++;
               break;
           case ...:
            ...
       }
       if (rs.maxLabAllocs() == rs.tlabAllocs()) {
           counters[TLAB_ALLOCS]++;
       } else if (rs.maxLabAllocs() == ...) {
          ...
        } else if ...
   
        // Process and update any other counters
        ...
}
```
The other idea is to do the global counter updates as they happen, but may be that isn't how visualizer works and the total counts aren't needed except upon request or at specific points in the visualization.

Since I am not familiar with the code or the visualizer, treat these comments with the appropriate dose of salt, and do get a review from someone familiar with the tool and the code.

Reply:
I think the main reason I have multiple Counters is because I tried to integrate it with the legend panel, which if you remember would be the column on the right that shows which color means what state. In that part of the code, it is written in a hashmap. Therefore, it is easier for me to separate all the counters so it can match up with the keys.

Reply:
I understand; I was suggesting that the counters could be updated via one sweep over the RegionStats, but I see that this is how the code was before, so what you are doing here isn't new. It just seemed a tad wasteful depending on how these counters are used, and perhaps not scalable as number of counters increases, especially as the heaps become bigger (but I guess a mitigating factor is that Shenandoah grows region sizes with heap, so the number of Regions doesn't grow forever). Probably fine for now.

Reply:
Okay that sounds good! I'll discuss more about this with William once he gets back.

Reply:
I like @ysramakrishna 's idea. The pattern of computing the snapshot properties upon request made more sense when there was only a few of them, but now we have _many_. It would be better to compute all the properties in a single pass over the region data. We also know that these properties will be rendered, so the lazy approach isn't necessary.

Reply:
I did create a bigger counter. However, I am not quite sure how I can integrate that with the legend panel so that it can put the data in the Hashmap so the visualizer can print it out.

Reply:
I just push the latest version of counter that works and doesn't count each time when I called the method. Maybe you can review that one.